taichi/program/state_flow_graph.cpp

```diff
@@ -61,33 +61,38 @@ using StateToNodesMap = StateFlowGraph::StateToNodesMap;
 // TODO: rename state to edge since we have not only state flow edges but also
 // dependency edges.
 
-StateToNodesMap::StateToNodesMap() {
-  // TODO(mask): need to resize |data_| to |kNumInlined| first?
-}
 void StateToNodesMap::clear() {
   sorted_ = false;
+  mask_ = 0;
   // It seems that once we are beyond the small buffer threshold, there's no
   // going back...
   data_.clear();
 }
 
 bool StateToNodesMap::has_edge(const Edge &e) const {
-  // TODO(mask): Could be a bit tricky once mask is implemented: we need to
-  // support both sorted and unsorted cases.
-  // Alternatively, only use this in sorted case, and don't call this inside
-  // insert_edge()?
   TI_ASSERT(sorted_);
   return std::binary_search(data_.begin(), data_.end(), e);
 }
 
+bool StateToNodesMap::has_state(const AsyncState &as) const {
+  for (auto iter = get_state_iterator(); !iter.done(); ++iter) {
+    if (iter.get_state() == as) {
+      return true;
+    }
+  }
+  return false;
+}
+
 void StateToNodesMap::insert_edge(const AsyncState &as, Node *n) {
   TI_ASSERT(!sorted_);
   const Edge e = std::make_pair(as, n);
-  if (std::find(data_.begin(), data_.end(), e) != data_.end()) {
+  const auto slot = get_mask_slot(e);
+  const MaskType mask_bit = (1ULL << slot);
```

Comment:
nit: I suggest `const auto mask_bit = (MaskType(1) << slot);` so that it's easier to switch `MaskType` to a custom class.

Reply:
good point, done